Loop/Managers/SleepStore.swift

```diff
@@ -0,0 +1,121 @@
+            with: .equalTo,
+            value: HKCategoryValueSleepAnalysis.asleep.rawValue
+        )
+        
+        getAverageSleepStartTime(matching: inBedPredicate, sampleLimit: sampleLimit) {
+            (result) in
+            switch result {
+            case .success(_):
+                completion(result)
+            case .failure(let error):
+                switch error {
+                case SleepStoreError.noSleepDataAvailable:
+                    // if there were no .inBed samples, check if there are any .asleep samples that could be used to estimate bedtime
+                    self.getAverageSleepStartTime(matching: asleepPredicate, sampleLimit: self.sampleLimit, completion)
+                default:
+                    // otherwise, call completion
+                    completion(result)
+                }
+            }
+            
+        }
+    }
+
+    fileprivate func getAverageSleepStartTime(matching predicate: NSPredicate, sampleLimit: Int, _ completion: @escaping (_ result: SleepStoreResult<Date>) -> Void) {
+        let sleepType = HKObjectType.categoryType(forIdentifier: HKCategoryTypeIdentifier.sleepAnalysis)!
+        
+        // get more-recent values first
+        let sortByDate = NSSortDescriptor(key: HKSampleSortIdentifierStartDate, ascending: false)
+        
+        let query = HKSampleQuery(sampleType: sleepType, predicate: predicate, limit: sampleLimit, sortDescriptors: [sortByDate]) { (query, samples, error) in
+
+            if let error = error {
+                completion(.failure(error))
+            } else if let samples = samples as? [HKCategorySample] {
+                guard !samples.isEmpty else {
+                    completion(.failure(SleepStoreError.noSleepDataAvailable))
+                    return
+                }
+                
+                // find the average hour and minute components from the sleep start times
+                let average = samples.reduce(0, {$0 + $1.startDate.timeOfDayInSeconds()}) / samples.count
+                let averageHour = average / 3600
+                let averageMinute = average % 3600 / 60
+                
+                // find the next time that the user will go to bed, based on the averages we've computed
+                if let bedtime = Calendar.current.nextDate(after: Date(), matching: DateComponents(hour: averageHour, minute: averageMinute), matchingPolicy: .nextTime), bedtime.timeIntervalSinceNow <= .hours(24) {
+                    completion(.success(bedtime))
+                } else {
+                    completion(.failure(SleepStoreError.noMatchingBedtime))
+                }
+            } else {
+                completion(.failure(SleepStoreError.unknownReturnConfiguration))
+            }
+        }
+        healthStore.execute(query)
+    }
+}
+
+extension Date {
+    fileprivate func timeOfDayInSeconds() -> Int {
+        let calendar = Calendar.current
```

Comment:
One detail here I forgot to mention—the time of day for a particular sleep sample is dependent upon the time zone in which the sample was recorded. For example, if a user went on vacation recently, using their home time zone to determine the time of day in seconds could yield a result that doesn't accurately reflect when they went to sleep in their _vacation location's_ time zone.

Add a parameter to this function of type TimeZone, and assign it to the `timeZone` property of `calendar`. You can retrieve the TimeZone for each sleep sample by retrieving the name string from the sample's metadata via [HKMetadataKeyTimeZone](https://developer.apple.com/documentation/healthkit/hkmetadatakeytimezone).

Reply:
For unwrapping the timezone, what could an appropriate default be? We could use the current time zone, or exclude the sample from the analysis (though this would probably mean it wouldn't be possible to use a reduce function to average the samples)

Reply:
I just checked my sleep data, from SleepTracker, and it has timezone. Whether or not this field is present will depend on the app, and since it's metadata, the app isn't forced to store it. I think assuming local timezone in that case is probably the best option, as the user probably will have no samples with timezone.

Reply:
Sounds good. Just checked my sleep data (some from Bedtime, some from SleepCycle) and it all contains timezones